### PR TITLE
Add course statistic routes

### LINKED ISSUE #460: Overworld Backend: create endpoint for useful course statistics
## Current Behavior
Currently, no statistic endpoint exists for course statistics.

## Expected Behavior
There is an endpoint with useful statistic endpoints which provide good information about the course.

## DoD
- [x] decide what statistics to implement
- [x] an endpoint exists to get this data
- [x] unit tests are written

### PR DESCRIPTION
**Pull request by Aaron2000119**
Closes https://github.com/Gamify-IT/issues/issues/460.

Test with https://github.com/Gamify-IT/lecturer-interface/pull/53.

Adds the endpoints for course statistics to the overworld backend.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -97,6 +97,11 @@
 			<version>2.7.3</version>
 			<scope>compile</scope>
 		</dependency>
+		<dependency>
+			<groupId>com.fasterxml.jackson.datatype</groupId>
+			<artifactId>jackson-datatype-jsr310</artifactId>
+			<version>2.14.2</version>
```
Comment (delvh) [inline pom.xml]: 1. Why `2.14.2` when `2.15.0` exists?
2. Please add the new package to the third party licenses.

File: src/test/java/de/unistuttgart/overworldbackend/PlayerStatisticControllerTest.java [hunk 2]
```diff
@@ -158,6 +158,7 @@ public void createBasicData() {
         playerstatistic.setUnlockedAreas(unlockedAreas);
         playerstatistic.setCompletedDungeons(new ArrayList<>());
         initialPlayerStatistic = playerstatisticRepository.save(playerstatistic);
+        System.out.println(initialPlayerStatistic.getDate());
```
Comment (delvh) [inline src/test/java/de/unistuttgart/overworldbackend/PlayerStatisticControllerTest.java]:
```suggestion
```

File: src/main/java/de/unistuttgart/overworldbackend/data/statistics/LastPlayed.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+import lombok.NoArgsConstructor;
+import lombok.experimental.FieldDefaults;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@FieldDefaults(level = AccessLevel.PRIVATE)
+public class LastPlayed {
+
+    int hour;
```
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/data/statistics/LastPlayed.java:15]: `Instant hour`?

File: src/main/java/de/unistuttgart/overworldbackend/data/statistics/ActivePlayersPlaytime.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package de.unistuttgart.overworldbackend.data.statistics;
+
+public class ActivePlayersPlaytime {
+
+    int players;
+    int hour;
```
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/data/statistics/ActivePlayersPlaytime.java]: `Duration hours`?

File: src/main/java/de/unistuttgart/overworldbackend/data/PlayerStatisticDTO.java [hunk 5]
```diff
@@ -34,5 +35,9 @@ public class PlayerStatisticDTO {
     String userId;
     String username;
 
+    LocalDateTime date;
```
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/data/PlayerStatisticDTO.java]: I don't quite understand, what's the purpose of this second date?

File: src/main/java/de/unistuttgart/overworldbackend/controller/CourseStatisticController.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+@RestController
+@Slf4j
+@RequestMapping("/courses/{courseId}/statistics")
+public class CourseStatisticController {
+
+    @Autowired
+    JWTValidatorService jwtValidatorService;
+
+    @Autowired
+    CourseStatisticService courseStatisticService;
```
Comment (maexled) [inline src/main/java/de/unistuttgart/overworldbackend/controller/CourseStatisticController.java]:
```suggestion
    private JWTValidatorService jwtValidatorService;

    @Autowired
    private CourseStatisticService courseStatisticService;
```

File: src/main/java/de/unistuttgart/overworldbackend/data/statistics/ActivePlayersPlaytime.java [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package de.unistuttgart.overworldbackend.data.statistics;
+
+public class ActivePlayersPlaytime {
```
Comment (maexled) [inline src/main/java/de/unistuttgart/overworldbackend/data/statistics/ActivePlayersPlaytime.java]:
```suggestion
@Data
@NoArgsConstructor
@AllArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE)
public class ActivePlayersPlaytime {
```

File: src/main/java/de/unistuttgart/overworldbackend/service/CourseStatisticService.java [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@Transactional
+@Slf4j
+@FieldDefaults(level = AccessLevel.PRIVATE)
+public class CourseStatisticService {
+
+    static final List<Integer> LAST_PLAYED_RANGE = List.of(1, 3, 12, 24, 3 * 24, 7 * 24, 14 * 24);
```
Comment (maexled) [inline src/main/java/de/unistuttgart/overworldbackend/service/CourseStatisticService.java:25]:
```suggestion
    // time in hours
    static final List<Integer> LAST_PLAYED_RANGE = List.of(1, 3, 12, 24, 3 * 24, 7 * 24, 14 * 24);
```

File: src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java [hunk 9]
```diff
@@ -254,6 +253,14 @@ public void checkForUnlockedAreas(final Area currentArea, final PlayerStatistic
         }
     }
 
+    public PlayerStatisticDTO setActive(final int CourseId, final String playerId) {
+        final PlayerStatistic playerStatistic = getPlayerStatisticFromCourse(CourseId, playerId);
+        playerStatistic.setLastActive(LocalDateTime.now());
+        return playerstatisticMapper.playerStatisticToPlayerstatisticDTO(
+            playerstatisticRepository.save(playerStatistic)
+        );
+    }
```
Comment (maexled) [inline src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java]:
```suggestion
    public PlayerStatisticDTO setActive(final int courseId, final String playerId) {
        final PlayerStatistic playerStatistic = getPlayerStatisticFromCourse(courseId, playerId);
        playerStatistic.setLastActive(LocalDateTime.now());
        return playerstatisticMapper.playerStatisticToPlayerstatisticDTO(
            playerstatisticRepository.save(playerStatistic)
        );
    }
```

Comment (maexled) [reply src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java]: Also JavaDoc missing

File: src/main/java/de/unistuttgart/overworldbackend/service/CourseStatisticService.java [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+
+    private static void addLastPlayed(final PlayerStatistic playerStatistic, final List<LastPlayed> lastPlayed) {
+        lastPlayed
+            .parallelStream()
+            .filter(lastPlayedStatistic -> isInRange(lastPlayedStatistic, playerStatistic.getDate()))
+            .findFirst()
+            .ifPresent(lastPlayedStatistic -> lastPlayedStatistic.setPlayers(lastPlayedStatistic.getPlayers() + 1));
+    }
+
+    private static boolean isInRange(final LastPlayed lastPlayed, final LocalDateTime date) {
```
Comment (maexled) [inline src/main/java/de/unistuttgart/overworldbackend/service/CourseStatisticService.java:172]: add a little JavaDoc

File: src/main/java/de/unistuttgart/overworldbackend/service/CourseStatisticService.java [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+                    if (i == 0) {
+                        return true;
+                    } else return LAST_PLAYED_RANGE.get(i - 1) * 60L > minutes;
+                }
+            }
+        } else return lastPlayed.getHour() == LAST_PLAYED_RANGE.get(LAST_PLAYED_RANGE.size() - 1);
+        return false;
+    }
+
+    private static void addPlayerJoined(
```
Comment (maexled) [inline src/main/java/de/unistuttgart/overworldbackend/service/CourseStatisticService.java:191]: also a little JavaDoc please

File: src/test/java/de/unistuttgart/overworldbackend/MinigameTaskStatisticTest.java [hunk 12]
```diff
@@ -105,13 +105,10 @@ public void createBasicData() {
...
         world.setConfigured(true);
 
-        final Course course = new Course(
-            "PSE",
-            "SS-22",
-            "Basic lecture of computer science students",
-            true,
-            List.of(world)
-        );
+        List<World> worlds = new ArrayList<>();
```
Comment (maexled) [inline src/test/java/de/unistuttgart/overworldbackend/MinigameTaskStatisticTest.java]:
```suggestion
        final List<World> worlds = new ArrayList<>();
```

File: src/test/java/de/unistuttgart/overworldbackend/PlayerStatisticControllerTest.java (hunk 2 above)
Comment (maexled) [inline src/test/java/de/unistuttgart/overworldbackend/PlayerStatisticControllerTest.java]:
```suggestion
```

**Review by maexled — APPROVED** [review]

**Review by Gr33ndev — APPROVED** [review]

File: src/main/java/de/unistuttgart/overworldbackend/service/CourseStatisticService.java [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    private String getAreaName(final Area maxArea) {
+        final String name;
+        if (maxArea instanceof Dungeon dungeon) {
+            name = dungeon.getWorld().getStaticName() + " - " + dungeon.getStaticName();
+        } else {
+            name = maxArea.getStaticName();
+        }
+        return name;
```
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/CourseStatisticService.java]:
```suggestion
        if (maxArea instanceof Dungeon dungeon) {
            return dungeon.getWorld().getStaticName() + " - " + dungeon.getStaticName();
        } else {
            return maxArea.getStaticName();
        }
```

File: src/main/java/de/unistuttgart/overworldbackend/data/PlayerStatisticDTO.java (hunk 5 above)
Comment (Aaron2000119) [reply src/main/java/de/unistuttgart/overworldbackend/data/PlayerStatisticDTO.java]: one date is the creation date and the other date is the date the player last played

File: src/main/java/de/unistuttgart/overworldbackend/data/statistics/LastPlayed.java (hunk 3 above)
Comment (Aaron2000119) [reply src/main/java/de/unistuttgart/overworldbackend/data/statistics/LastPlayed.java:15]: i simply choose the same name as the frontend, which was developed first

File: src/main/java/de/unistuttgart/overworldbackend/data/PlayerStatisticDTO.java (hunk 5 above)
Comment (delvh) [reply src/main/java/de/unistuttgart/overworldbackend/data/PlayerStatisticDTO.java]: Shouldn't we call it `created` then?

**Review by delvh — APPROVED** [review]

**Comment by delvh** [comment]
It's ready to merge, together with the lecturer interface PR.
@Aaron2000119 or @maexled, do you want to merge, or should I do it?

**Pull request merged**